### PR TITLE
Resolution choose fix and improvement

### PR DESCRIPTION
**Pull request by Karlson2k**
This is an improvement for https://github.com/xbmc/xbmc/pull/19124 by @fritsch.

The [first commit](d7a06b63c680229fc46d5f8a73cf4f15c3a91051) allow to find ideal fit resolution in double framerate resolutions (if allowed by settings) even if something was found in single framerate match.

The [second commit](b140f7fdc39ee7a65f00c8b1e30357d409e59207) basically improve choosing of the best match by `fit percentage` instead of absolute value. Also it contains some optimisation (return early if ideal match is found, better logging) and applied the same algorithm for 3:2 pulldown resolutions as well.

It's up to @DaveTBlake whether to use in for v19 or for v20, but I would recommend to take at least the [first commit](d7a06b63c680229fc46d5f8a73cf4f15c3a91051) for v19.

### CODE REVIEW

File: xbmc/windowing/Resolution.cpp [hunk 1]
```diff
@@ -140,20 +139,28 @@ void CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, int hei
         (info.dwFlags & D3DPRESENTFLAG_MODEMASK) == (curr.dwFlags & D3DPRESENTFLAG_MODEMASK) &&
         MathUtils::FloatEquals(info.fRefreshRate, fps, 0.01f))
     {
+      float fit_ratio = (height == info.iScreenHeight)
+                             ? static_cast<float>(width) / static_cast<float>(info.iScreenWidth)
+                             : static_cast<float>(height) / static_cast<float>(info.iScreenHeight);
+      if (fit_ratio > 1.0f)
+        fit_ratio = 1.0f / fit_ratio;
+
```
Comment (fritsch) [inline xbmc/windowing/Resolution.cpp:147]: Keep in mind, that many movies are not 1920x1080 - they are only 1920x800 or 1920x816 as users rip it that way to save the "black bars".

So in many cases height won't be iScreenHeight at all. The aspect ratio should be extracted from the movie itself.

File: xbmc/windowing/Resolution.cpp [hunk 2]
```diff
@@ -211,14 +225,28 @@ void CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, int hei
...
-            "[WHITELIST] Matched an exact resolution with a 3:2 pulldown refresh rate {} ({})",
-            info.strMode, i);
-        resolution = i;
-        return;
+        float fit_ratio = (height == info.iScreenHeight)
+                              ? static_cast<float>(width) / static_cast<float>(info.iScreenWidth)
+                              : static_cast<float>(height) / static_cast<float>(info.iScreenHeight);
+        if (fit_ratio > 1.0f)
+          fit_ratio = 1.0f / fit_ratio;
+
```
Comment (fritsch) [inline xbmc/windowing/Resolution.cpp:233]: That's the duplication from above. It won't work in most cases. AR needs to come from videoplayer.

File: xbmc/windowing/Resolution.cpp (hunk 1 above)
Comment (Karlson2k) [reply xbmc/windowing/Resolution.cpp:147]: @fritsch, it's not about **aspect** ratio. It is used percentage of the output resolution, i.e. percentage video use in output (or 100% - percent of "black bars").

File: xbmc/windowing/Resolution.cpp (hunk 2 above)
Comment (Karlson2k) [reply xbmc/windowing/Resolution.cpp:233]: The same. It is not about AR.

**Comment by Karlson2k** [comment]
Maybe I should rename variable to void confusion.

Examples:
I have monitor 16:9 monitor with native resolution 2560x1440, it supports 1920x1080 and 1920x1200 signals.
For video 1920x800 fit ratio: for 1920x1080 is 0.7407 and for 1920x1200 is 0.6667. 1920x1080 is better and will be used.
For video 1920x1080 fit ratio: for 1920x1080 is 1.0 and for 1920x1200 is 0.9. 1920x1080 is better and will be used.
For (hypothetical) video 1920x1150 fit ratio: for 1920x1080 - does not fit and not be checked, for 1920x1200 is 0,9583. 1920x1200 will be used.

For the case from issue https://github.com/xbmc/xbmc/issues/19037:
Video 3840x2160 fit ratio: for 4096x2160 is 0.9375, for 3840x2160 is 1.0. 3840x2160 will be used.

**Comment by Karlson2k** [comment]
PR was updated, variables renamed.
For clarity: **the current PR do nothing with aspect ratio (AR). It checks match only by resolution.**

However we still have some space to improve: we should check native screen AR and choose match by height only or by width only, depending of video AR. 

Examples:
Screen is 16:9. 
For videos 16:9 we should match by height or width (doesn't matter).
For videos 21:9 we should match by width.
For videos 4:3 we should match by height.
This way we can avoid double "black bars", vertical and horizontal at the same time.

**Comment by fritsch** [comment]
Hi,

your PR consists of two parts, let's start with the percentage metrics.
It's equivalent to the code we already have, but only in case where height exactly matches. If it does not exactly match, it by default uses the else case and only the width ratio is considered. The resulting formula is only takes a percentage ratio distance (0 to 1.0) of the dimension into account which does not fully match, e.g. if height matches it chooses with, if height does not match, it uses the height ratio. Minkowski does a similar thing, therefore if the height matches that part of the addition is just 0 while the width difference will be summed up. So for the height equals screenHeight they are identical. Yours converges to 1.0 for a perfect ratio, my distance converges to 0, which is a nice thing - as the nearer to zero. Small details.

See below example:

Let's choose your example 1920 x 800 is matched to 1920 x 1080 (1) and 1920 x 1200 (2)
My code:
Match to 1: 0 + 280 = 280
Match to 2: 0 + 400 = 400

1 is chosen.

Your code:
(1) 1080 / 800 = 1,35 -> inverse: 0,74071
(2) 1200 / 800 = 1,5 -> inverse: 0.667

Basically I see no difference here from the achievable results _in_ the given constraints, e.g. width and height, at least one is nicely matching. There are very heavy difference if it would be used with arbitrary dimensions as in your calculation only one dimension's ratio plays a role.

2) You add the same algorithm to the Pull-Down 3:2 Modes
This makes sense, but can be equally achieved with Minkowski

Do you agree?

**Comment by Karlson2k** [comment]
Thank for review, @fritsch!

Let's start from your second point.
First of all you missed my [first commit](https://github.com/xbmc/xbmc/pull/d7a06b63c680229fc46d5f8a73cf4f15c3a91051). Currently if output supports 4096x2160 30Hz 60Hz and 3840x2160 60Hz only, for video 3840x2160 30fps code selects  4096x2160 30Hz as best matching.
My [first commit](https://github.com/xbmc/xbmc/pull/d7a06b63c680229fc46d5f8a73cf4f15c3a91051) fixes it and allow to select 3840x2160 60Hz as output in such case.

> You add the same algorithm to the Pull-Down 3:2 Modes
This makes sense, but can be equally achieved with Minkowski

Yes, 3:2 can be implemented with your algorithm, but I'll explain why my algorithm is better.

So, for your first point.
Let's take as an example video 1920x1080. And the output doesn't support 1920x1080 for any reason (or it isn't whitelisted), but supports 2048x1080 and 1920x1200.
With your implementation: 
For 2048x1080 penalty is `128`
For 1920x1200 penalty is `120`. `120` is lower than `128`, 1920x1200 will be chosen as the best one.
With my implementation:
For 2048x1080 fit_percentage is 93.75% (0.9375)
For 1920x1200 fit_percentage is 90% (0.9)
`93.75%` is better than `90%` so 1920x1200 will be selected.
And that is correct, because for 2048x1080 video will use (rounded) 94% of surface area, while for 1920x1200 video will use only 90% of surface area.
That's why my idea with relative values is better than absolute values.

> It's equivalent to the code we already have, but only in case where height exactly matches.

Actually we always have one of the dimensions precisely matched (otherwise why it is "native"). So your code always use **either** height difference **or** width difference for every output resolution.
The only difference that your version compares absolute value, while my version compare relative value.

Simplified example: for 21:9 video 100 pixels difference in width is less notable than 100 pixels difference in height.
You can increase weight of height difference but it will be equivalent to my algo.

**Comment by fritsch** [comment]
> Let's take as an example video 1920x1080. And the output doesn't support 1920x1080 for any reason (or it isn't whitelisted), but supports 2048x1080 and 1920x1200.

What is 2048x1080? :-) What about 1928 x 1072, which would eventually nicely scale into 1920 x 1080?

Come on. 

But you exactly prove the point I made .. I fixed the low hanging fruits for v19. While I only chose the resolution where a 1:1 pixel mapping is very likely. As I said before I did not have interest at all to best-fit into double scaling problems.

I think for this PR you need someone else to review it, as we come from metric comparison to which resolution match comparison - and this is - as now said many times - exactly what I wanted to avoid.

**Comment by FernetMenta** [comment]
You revert the original idea of keep it simple. It was born through the findings that there is no one-fits-all algo that satisfies every single scenario. Users always have the chance of manually selecting a resolution of their choice.

**Comment by Karlson2k** [comment]
> > Let's take as an example video 1920x1080. And the output doesn't support 1920x1080 for any reason (or it isn't whitelisted), but supports 2048x1080 and 1920x1200.
> 
> What is 2048x1080? :-) 

It is DCI 2K resolution. A pretty standard one. https://en.wikipedia.org/wiki/2K_resolution
My example was a real-life example.

> What about 1928 x 1072, which would eventually nicely scale into 1920 x 1080?

What??? How 1928x1072 could be mapped to 1920x1080?
It could be mapped to 1920x1080 only if 1920x1080 and desktop resolution.
So we can do just nothing with 1928 x 1072. Current code doesn't match it for anything.

> Come on.

Exactly. :)


> But you exactly prove the point I made .. I fixed the low hanging fruits for v19. While I only chose the resolution where a 1:1 pixel mapping is very likely. As I said before I did not have interest at all to best-fit into double scaling problems.
> 
> I think for this PR you need someone else to review it, as we come from metric comparison to which resolution match comparison - and this is - as now said many times - exactly what I wanted to avoid.

I'm fine if we skip it for v19, while I really don't think that my implementation is more complicated in any way.

I'll make another PR with fixes from this PR, that you skip every every time.

**Comment by Karlson2k** [comment]
> You revert the original idea of keep it simple. It was born through the findings that there is no one-fits-all algo that satisfies every single scenario. Users always have the chance of manually selecting a resolution of their choice.

Do you think that best match by (x1 - x2) is much simpler compared to best match by (x1/x2)?
From my point of view they are equal in simplicity.

**Comment by DaveTBlake** [comment]
>I'm fine if we skip it for v19, while I really don't think that my implementation is more complicated in any way.

Well I'm well confused and so this close to release I'm going to label v20. If  you guys  then agree something for v19, that is fine.

**Comment by ilikenwf** [comment]
Sorry to bump an old PR and issue, but any chance this will get rebased? I am running into an issue where if I have the UI set to 3840x2160p, and all valid modes whitelisted, that playing a 1920x816 encode (where the video is "cinematic" aspect ratio) results in Kodi incorrectly using 3840x2160 as the resolution for playback, bypassing upscaling by my TV.

**Comment by Karlson2k** [comment]
This needs a re-write.
In my TODO :)
